snuba/api.py

```diff
@@ -162,16 +162,80 @@ def query(validated_body=None, timer=None):
+    """Generate a SQL string from the parameters."""
+    aggregate_exprs = [util.column_expr(dataset, col, body, alias, agg) for (agg, col, alias) in body['aggregations']]
+    groupby = util.to_list(body['groupby'])
+    group_exprs = [util.column_expr(dataset, gb, body) for gb in groupby]
+    selected_cols = [util.column_expr(dataset, util.tuplify(colname), body) for colname in body.get('selected_columns', [])]
+    select_clause = u'SELECT {}'.format(', '.join(group_exprs + aggregate_exprs + selected_cols))
+
+    from_clause = u'FROM {}'.format(table)
+    if final:
+        from_clause = u'{} FINAL'.format(from_clause)
+    if 'sample' in body:
+        from_clause = u'{} SAMPLE {}'.format(from_clause, body['sample'])
+
+    join_clause = ''
+    if 'arrayjoin' in body:
+        join_clause = u'ARRAY JOIN {}'.format(body['arrayjoin'])
+
+    where_clause = ''
+    if body['conditions']:
+        where_clause = u'WHERE {}'.format(util.conditions_expr(dataset, body['conditions'], body))
+
+    prewhere_clause = ''
+    if prewhere_conditions:
+        prewhere_clause = u'PREWHERE {}'.format(util.conditions_expr(dataset, prewhere_conditions, body))
+
+    group_clause = ''
+    if groupby:
+        group_clause = 'GROUP BY ({})'.format(', '.join(util.column_expr(dataset, gb, body) for gb in groupby))
+        if body.get('totals', False):
+            group_clause = '{} WITH TOTALS'.format(group_clause)
+
+    having_clause = ''
+    having_conditions = body.get('having', [])
+    if having_conditions:
+        assert groupby, 'found HAVING clause with no GROUP BY'
+        having_clause = u'HAVING {}'.format(util.conditions_expr(dataset, having_conditions, body))
+
+    order_clause = ''
+    if body.get('orderby'):
+        orderby = [util.column_expr(dataset, util.tuplify(ob), body) for ob in util.to_list(body['orderby'])]
+        orderby = [u'{} {}'.format(ob.lstrip('-'), 'DESC' if ob.startswith('-') else 'ASC') for ob in orderby]
+        order_clause = u'ORDER BY {}'.format(', '.join(orderby))
+
+    limitby_clause = ''
+    if 'limitby' in body:
+        limitby_clause = 'LIMIT {} BY {}'.format(*body['limitby'])
+
+    limit_clause = ''
+    if 'limit' in body:
+        limit_clause = 'LIMIT {}, {}'.format(body.get('offset', 0), body['limit'])
+
+    return ' '.join([c for c in [
+        select_clause,
+        from_clause,
+        join_clause,
+        prewhere_clause,
+        where_clause,
+        group_clause,
+        having_clause,
+        order_clause,
+        limitby_clause,
+        limit_clause
+    ] if c])
+
+
 @split_query
 def parse_and_run_query(dataset, validated_body, timer):
     body = deepcopy(validated_body)
 
-    table = dataset.get_schema().get_table_name()
-
     max_days, date_align, config_sample = state.get_configs([
         ('max_days', None),
         ('date_align_seconds', 1),
-        ('sample', 1),
+        ('sample', None),
```

Comment:
Does this actually change anything or it is just to avoid the SAMPLE clause at all ?

Reply:
This will make it so that the `SAMPLE` clause is only added to the SQL if:

1. There is a user-defined sample rate (via `sample` or via `turbo` properties.)
2. There is not a user-defined sample rate but there is a configured sample rate.

I'm not sure there is much value in the globally configured sample rate, as noted below. I think we should consider removing it.